File: back/appointment-service/internal/interval_test.go
package common

import (
	"runtime"
	"slices"
	"testing"
	"time"
)

func TestInterval(t *testing.T) {
	start := time.Date(2020, 1, 1, 10, 0, 0, 0, time.UTC)
	interval := Interval{Start: start, End: start.Add(time.Hour)}

	if !interval.IsOverlap(interval) {
		t.Fatalf("interval is not overlapping itself %v", interval)
	}

	if !interval.IsFit(interval) {
		t.Fatalf("interval is not fitting itself %v", interval)
	}

	interval2 := Interval{Start: interval.End, End: interval.End.Add(time.Second)}
	if interval.IsOverlap(interval2) {
		t.Fatalf("intervals should not overlap %v %v", interval, interval2)
	}

	if interval.IsFit(interval2) {
		t.Fatalf("intervals should not fit %v %v", interval, interval2)
	}

	interval2.Start = interval2.Start.Add(-2 * time.Second)
	if !interval.IsOverlap(interval2) {
		t.Fatalf("intervals should not overlap %v %v", interval, interval2)
	}

	if interval.IsFit(interval2) {
		t.Fatalf("intervals should not fit %v %v", interval, interval2)
	}

	interval2.Start = interval.Start.Add(1 * time.Second)
	interval2.End = interval.End.Add(-1 * time.Second)
	if !interval.IsOverlap(interval2) {
		t.Fatalf("intervals is not overlapping %v %v", interval, interval2)
	}

	if !interval.IsFit(interval2) {
		t.Fatalf("intervals is not fitting %v %v", interval, interval2)
	}

	if !interval2.IsOverlap(interval) {
		t.Fatalf("intervals is not overlapping %v %v", interval, interval2)
	}

	if interval2.IsFit(interval) {
		t.Fatalf("intervals is fitting %v %v", interval, interval2)
	}

	interval2.Start = interval.Start
	interval2.End = interval.End.Add(-1 * time.Second)
	if !interval.IsOverlap(interval2) {
		t.Fatalf("intervals is not overlapping %v %v", interval, interval2)
	}

	interval = Interval{Start: start, End: start.Add(time.Hour)}
	result := interval.Subtract(interval)
	if result != nil {
		t.Fatalf("result should be nil %v", result)
	}

	interval2 = Interval{Start: start.Add(30 * time.Minute), End: start.Add(time.Hour)}
	expected := Interval{Start: start, End: start.Add(30 * time.Minute)}
	result = interval.Subtract(interval2)
	if len(result) != 1 || result[0] != expected {
		t.Fatalf("result should be %v but got %v", expected, result)
	}

	interval2 = Interval{Start: start.Add(31 * time.Minute), End: start.Add(time.Hour + time.Minute)}
	expected = Interval{Start: start, End: start.Add(31 * time.Minute)}
	result = interval.Subtract(interval2)
	if len(result) != 1 || result[0] != expected {
		t.Fatalf("result should be %v but got %v", expected, result)
	}

	interval2 = Interval{Start: start, End: start.Add(30 * time.Minute)}
	expected = Interval{Start: start.Add(30 * time.Minute), End: start.Add(time.Hour)}
	result = interval.Subtract(interval2)
	if len(result) != 1 || result[0] != expected {
		t.Fatalf("result should be %v but got %v", expected, result)
	}

	interval2 = Interval{Start: start.Add(-1 * time.Minute), End: start.Add(31 * time.Minute)}
	expected = Interval{Start: start.Add(31 * time.Minute), End: start.Add(time.Hour)}
	result = interval.Subtract(interval2)
	if len(result) != 1 || result[0] != expected {
		t.Fatalf("result should be %v but got %v", expected, result)
	}

	interval2 = Interval{Start: start.Add(10 * time.Minute), End: start.Add(time.Hour - 10*time.Minute)}
	expected2 := Intervals{
		{Start: start, End: start.Add(10 * time.Minute)},
		{Start: start.Add(time.Hour - 10*time.Minute), End: start.Add(time.Hour)},
	}
	result = interval.Subtract(interval2)
	if len(result) != 2 || result[0] != expected2[0] || result[1] != expected2[1] {
		t.Fatalf("result should be %v but got %v", expected2, result)
	}
}

func TestIntervalSlices(t *testing.T) {
	start := time.Now()
	interval := Interval{Start: start, End: start.Add(time.Minute)}

	intervals := Intervals{interval, interval}
	intervals.SortByStart()
	if !intervals.IsSorted() {
		t.Fatalf("intervals should be sorted %v", intervals)
	}

	if !intervals.HasOverlaps() {
		t.Fatalf("intervals should be overlap %v", intervals)
	}

	intervals = Intervals{}
	for i := range 10 {
		tp := start.Add(time.Duration(i) * time.Minute)
		intervals = append(intervals, Interval{Start: tp, End: tp.Add(1 * time.Minute)})
	}
	intervals.SortByStart()
	if !intervals.IsSorted() {
		t.Fatalf("intervals should be sorted %v", intervals)
	}
	if intervals.HasOverlaps() {
		t.Fatalf("intervals should has overlap %v", intervals)
	}

	intervals = append(intervals, intervals[0])
	if intervals.IsSorted() {
		t.Fatalf("intervals should not be sorted %v", intervals)
	}
	intervals.SortByStart()
	if !intervals.IsSorted() {
		t.Fatalf("intervals should not be sorted %v", intervals)
	}
	if !intervals.HasOverlaps() {
		t.Fatalf("intervals should has overlap %v", intervals)
	}
}

func TestIntervalsUnite(t *testing.T) {
	type Expected = Intervals
	checkCase := func(intervals Intervals, expected Expected) {
		_, filename, line, _ := runtime.Caller(1)
		intervals.Unite()
		if !slices.Equal(intervals, expected) {
			t.Fatalf("%s:%d: expected %v, got %v", filename, line, expected, intervals)
		}
	}

	// Test empty intervals
	checkCase(Intervals{}, Expected{})

	// Test single interval
	singleInterval := Intervals{
		{Start: time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 10, 0, 0, 0, time.UTC)},
	}
	checkCase(singleInterval, singleInterval.Copy())

	// Test non-overlapping intervals
	nonOverlapping := Intervals{
		{Start: time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 10, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 11, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 14, 0, 0, 0, time.UTC)},
	}
	checkCase(nonOverlapping, nonOverlapping.Copy())

	// Test overlapping intervals
	overlapping := Intervals{
		{Start: time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 11, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 10, 30, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 12, 1, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC)},
	}
	expectedOverlapping := Intervals{
		{Start: time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC)},
	}
	checkCase(overlapping, expectedOverlapping)

	// Test multiple overlapping intervals
	multipleOverlapping := Intervals{
		{Start: time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 11, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 10, 30, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 14, 1, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 14, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 15, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 14, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 15, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 15, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 16, 0, 0, 0, time.UTC)},
	}
	expectedMultipleOverlapping := Intervals{
		{Start: time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 15, 0, 0, 0, time.UTC)},
		{Start: time.Date(2024, 10, 9, 15, 0, 0, 0, time.UTC), End: time.Date(2024, 10, 9, 16, 0, 0, 0, time.UTC)},
	}
	checkCase(multipleOverlapping, expectedMultipleOverlapping)
}

func TestSetPassedIntervals(t *testing.T) {
	type Expected = Intervals
	checkCase := func(i Intervals, e Intervals, expected Expected) {
		result := i.PassedIntervals(e)
		if !slices.Equal(result, expected) {
			t.Fatalf("expected %v, got %v", expected, result)
		}
	}

	checkCase(
		Intervals{
			{
				time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
		},
		Intervals{
			{
				time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC)},
		},

		Expected{
			{
				time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC),
			}, {
				time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
		},
	)

	checkCase(
		Intervals{
			{
				time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
		},
		Intervals{
			{
				time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
		},

		Expected{
			{
				Start: time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC),
				End:   time.Date(2024, 10, 9, 12, 0, 0, 0, time.UTC),
			},
		},
	)

	checkCase(
		Intervals{
			{
				time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
		},
		Intervals{
			{
				time.Date(2024, 10, 9, 8, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
		},

		Expected{},
	)

	checkCase(
		Intervals{},
		Intervals{
			{
				time.Date(2024, 10, 9, 8, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
		},

		Expected{},
	)

	checkCase(
		Intervals{
			{
				time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},

			{
				time.Date(2024, 10, 9, 20, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 21, 0, 0, 0, time.UTC),
			},
		},
		Intervals{
			{
				time.Date(2024, 10, 9, 10, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 11, 0, 0, 0, time.UTC),
			},

			{
				time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 14, 0, 0, 0, time.UTC),
			},

			{
				time.Date(2024, 10, 9, 15, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 16, 0, 0, 0, time.UTC),
			},
		},

		Expected{
			{
				time.Date(2024, 10, 9, 9, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 10, 0, 0, 0, time.UTC),
			},
			{
				time.Date(2024, 10, 9, 11, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 13, 0, 0, 0, time.UTC),
			},
			{
				time.Date(2024, 10, 9, 14, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 15, 0, 0, 0, time.UTC),
			},
			{
				time.Date(2024, 10, 9, 16, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 18, 0, 0, 0, time.UTC),
			},
			{
				time.Date(2024, 10, 9, 20, 0, 0, 0, time.UTC),
				time.Date(2024, 10, 9, 21, 0, 0, 0, time.UTC),
			},
		},
	)

}

File: back/appointment-service-cpp/interval_test.cpp
#include <bits/chrono.h>
#include <gtest/gtest.h>

#include <chrono>
#include <vector>
import IntervalModule;

using Clock = std::chrono::system_clock;
using TimePoint = Clock::time_point;
using Interval = ::lf::Interval<TimePoint>;
using Intervals = ::lf::Intervals<TimePoint>;

using namespace std::chrono;

constinit TimePoint kStartPoint =
    sys_days{2020y / January / 01} + 10h;  // 2020-01-01 10:00:00 // UTC ?

TEST(IntervalTest, BasicOperations) {
  TimePoint start = kStartPoint;
  Interval interval{start, start + 1h};

  EXPECT_TRUE(interval.is_overlap(interval));
  EXPECT_TRUE(interval.is_fit(interval));

  Interval interval2{interval.end, interval.end + 1s};
  EXPECT_FALSE(interval.is_overlap(interval2));
  EXPECT_FALSE(interval.is_fit(interval2));

  interval2.start -= 2s;
  EXPECT_TRUE(interval.is_overlap(interval2));
  EXPECT_FALSE(interval.is_fit(interval2));

  interval2.start = interval.start + 1s;
  interval2.end = interval.end - 1s;
  EXPECT_TRUE(interval.is_overlap(interval2));
  EXPECT_TRUE(interval.is_fit(interval2));
  EXPECT_TRUE(interval2.is_overlap(interval));
  EXPECT_FALSE(interval2.is_fit(interval));

  interval2.start = interval.start + 1s;
  interval2.end = interval.end - 1s;
  EXPECT_TRUE(interval.is_overlap(interval2));
}

TEST(IntervalTest, Subtract) {
  TimePoint start = kStartPoint;

  Interval interval{start, start + 1h};
  auto result = interval.subtract(interval);
  EXPECT_TRUE(result.empty());

  // Interval interval2{start + 30min, start + 1h};
  Interval expected{start, start + 30min};
  result = interval.subtract({start + 30min, start + 1h});
  EXPECT_EQ(result.size(), 1);
  EXPECT_EQ(result[0], expected);

  // 76
  result = interval.subtract({start + 31min, start + 1h + 1min});
  expected = Interval{start, start + 31min};
  EXPECT_EQ(result.size(), 1);
  EXPECT_EQ(result[0], expected);

  // 83
  expected = Interval{start + 30min, start + 1h};
  result = interval.subtract({start, start + 30min});
  EXPECT_EQ(result.size(), 1);
  EXPECT_EQ(result[0], expected);
  // 90
  expected = Interval{start + 31min, start + 1h};
  result = interval.subtract({start - 1min, start + 31min});
  EXPECT_EQ(result.size(), 1);
  EXPECT_EQ(result[0], expected);
  // 97
  auto expected2 = Intervals{
      {start, start + 10min},
      {start + 1h - 10min, start + 1h},
  };
  result = interval.subtract({start + 10min, start + 1h - 10min});
  EXPECT_EQ(result.size(), 2);
  EXPECT_EQ(result[0], expected2[0]);
  EXPECT_EQ(result[1], expected2[1]);
}

TEST(IntervalTest, SortingAndOverlap) {
  Interval interval{kStartPoint, kStartPoint + 1min};

  Intervals intervals = {interval, interval};
  sort_by_start(intervals);
  EXPECT_TRUE(is_sorted(intervals));
  EXPECT_TRUE(has_overlaps(intervals));

  intervals.clear();
  for (int i = 0; i < 10; ++i) {
    auto tp = kStartPoint + i * 1min;
    intervals.push_back({tp, tp + 1min});
  }
  sort_by_start(intervals);
  EXPECT_TRUE(is_sorted(intervals));
  EXPECT_FALSE(has_overlaps(intervals));

  intervals.push_back(intervals[0]);
  EXPECT_FALSE(is_sorted(intervals));
  sort_by_start(intervals);
  EXPECT_TRUE(is_sorted(intervals));
  EXPECT_TRUE(has_overlaps(intervals));
}

TimePoint make_time(TimePoint::duration seconds) {
  return kStartPoint + seconds;
}

TEST(PrepareUnitedTests, MergesOverlappingIntervals) {
  Intervals intervals = {{make_time(0min), make_time(10min)},
                         {make_time(5min), make_time(15min)},
                         {make_time(12min), make_time(20min)},
                         {make_time(18min), make_time(25min)},
                         {make_time(30min), make_time(40min)},
                         {make_time(35min), make_time(50min)}};

  prepare_united(intervals);

  ASSERT_EQ(intervals.size(), 2);
  EXPECT_EQ(intervals[0].start, make_time(0min));
  EXPECT_EQ(intervals[0].end, make_time(25min));
  EXPECT_EQ(intervals[1].start, make_time(30min));
  EXPECT_EQ(intervals[1].end, make_time(50min));

  intervals = {{make_time(0min), make_time(10min)},
               {make_time(5min), make_time(15min)},
               {make_time(15min), make_time(20min)}};

  prepare_united(intervals);

  ASSERT_EQ(intervals.size(), 2);
  EXPECT_EQ(intervals[0].start, make_time(0min));
  EXPECT_EQ(intervals[0].end, make_time(15min));
  EXPECT_EQ(intervals[1].start, make_time(15min));
  EXPECT_EQ(intervals[1].end, make_time(20min));
}

TEST(PrepareUnitedTests, KeepsNonOverlappingIntervals) {
  Intervals intervals = {{make_time(0s), make_time(5s)},
                         {make_time(10s), make_time(15s)},
                         {make_time(20s), make_time(25s)},
                         {make_time(30s), make_time(35s)},
                         {make_time(40s), make_time(45s)}};

  auto copy = intervals;
  prepare_united(intervals);

  ASSERT_EQ(intervals.size(), 5);  // No merging should occur
  ASSERT_TRUE(std::ranges::equal(intervals, copy));

  intervals = {
      {make_time(0s), make_time(10s)},
      {make_time(10s), make_time(20s)},
      {make_time(20s), make_time(30s)},
      {make_time(30s), make_time(40s)},
  };

  copy = intervals;
  prepare_united(intervals);

  ASSERT_EQ(intervals.size(), 4);
  ASSERT_TRUE(std::ranges::equal(intervals, copy));
}

TEST(PrepareUnitedTests, RemainsUnchanged) {
  Intervals intervals;
  prepare_united(intervals);
  ASSERT_TRUE(intervals.empty());

  intervals = {{make_time(0s), make_time(10s)}};

  prepare_united(intervals);

  ASSERT_EQ(intervals.size(), 1);
  EXPECT_EQ(intervals[0].start, make_time(0s));
  EXPECT_EQ(intervals[0].end, make_time(10s));
}

TEST(PrepareUnitedTests, MixedMergingAndNonMerging) {
  Intervals intervals = {
      {make_time(0s), make_time(10s)},
      {make_time(5s), make_time(12s)},  // Overlaps with first
      {make_time(20s), make_time(25s)},
      {make_time(30s), make_time(35s)},
      {make_time(32s), make_time(40s)},  // Overlaps with previous
      {make_time(40s), make_time(60s)}};

  prepare_united(intervals);

  ASSERT_EQ(intervals.size(), 4);
  EXPECT_EQ(intervals[0].start, make_time(0s));
  EXPECT_EQ(intervals[0].end, make_time(12s));
  EXPECT_EQ(intervals[1].start, make_time(20s));
  EXPECT_EQ(intervals[1].end, make_time(25s));
  EXPECT_EQ(intervals[2].start, make_time(30s));
  EXPECT_EQ(intervals[2].end, make_time(40s));
  EXPECT_EQ(intervals[3].start, make_time(40s));
  EXPECT_EQ(intervals[3].end, make_time(60s));
}

TEST(IntervalsTest, TestSetPassedIntervals) {
  auto checkCase = [](const Intervals& i, const Intervals& e,
                      const Intervals& expected) {
    auto result = passed_intervals(i, e);
    EXPECT_TRUE(std::ranges::equal(result, expected))
        << "Expected: " << expected.size()
        << " intervals, got: " << result.size();
  };

  const auto start_time = sys_days{2024y / October / 9};
  // 157
  checkCase({{{start_time + 9h}, {start_time + 18h}}},
            {{{start_time + 12h}, {start_time + 13h}}},
            {{{start_time + 9h}, {start_time + 12h}},
             {{start_time + 13h}, {start_time + 18h}}});
  // 181
  checkCase({{{start_time + 9h}, {start_time + 18h}}},
            {{{start_time + 12h}, {start_time + 18h}}},
            {{{start_time + 9h}, {start_time + 12h}}});

  // 203
  checkCase(
      {
          {
              {start_time + 9h},
              {start_time + 18h},
          },
      },
      {
          {
              {start_time + 8h},
              {start_time + 18h},
          },
      },
      Intervals{});

  // 220
  checkCase({},
            {
                {
                    {start_time + 8h},
                    {start_time + 18h},
                },
            },
            {});

  // 232
  checkCase(
      {
          {
              start_time + 9h,
              start_time + 18h,
          },
          {
              start_time + 20h,
              start_time + 21h,
          },
      },
      {
          {
              start_time + 10h,
              start_time + 11h,
          },
          {
              start_time + 13h,
              start_time + 14h,
          },
          {
              start_time + 15h,
              start_time + 16h,
          },
      },
      {
          {
              start_time + 9h,
              start_time + 10h,
          },
          {
              start_time + 11h,
              start_time + 13h,
          },
          {
              start_time + 14h,
              start_time + 15h,
          },
          {
              start_time + 16h,
              start_time + 18h,
          },
          {
              start_time + 20h,
              start_time + 21h,
          },
      });
}

int main(int argc, char** argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}
